File: linked_list/singly_linked2.py
# https://www.geeksforgeeks.org/insertion-in-linked-list/

"""
	In this post, methods to insert a new node in the linked list are discussed. A node can be added in three ways 
    - At the front of the linked list  
    - After a given node. 
    - At the end of the linked list.

	Method by using a constructor call to reduce memory allocation work and code lines.

	Time Complexity: O(N) 
	Auxiliary Space: O(1)
"""

# Node class
class Node:

	# A constructor is called here
	def __init__(self, data):
		self.data = data # Automatically assign data
		self.next = None # Initialize next pointer as null


# Linked List class contains a Node object
class LinkedList:

	# Function to initialize head
	def __init__(self):
		self.head = None

	'''This function is in LinkedList class.
	It inserts a new node at the beginning'''
	def insertAtHead(self, new_data):

		# 1 & 2: Allocate the Node &
		#	 Put in the data
		new_node = Node(new_data)

		# 3. Make next of new Node as head
		new_node.next = self.head

		# 4. Move the head to point to new Node
		self.head = new_node

	'''This function is in LinkedList class. It inserts a
	new node after the given prev_data. This method is
	defined inside LinkedList class shown above'''
	def insertAfter(self, prev_data, new_data):

		# 1. check if the Linked List is empty or not
		if self.head is None:
			return

		# 2. Create new node & Put in the data
		new_node = Node(new_data)

		# 3. If prev_data is at the first position
		if (prev_data == self.head.data):
			new_node.next = self.head.next
			self.head.next = new_node
			return

		# 4. check if the given prev_data exists
		head = self.head
		while (head.data != prev_data):
			head = head.next
			if head == None:
				return
		new_node.next = head.next
		head.next = new_node

	'''This function is defined in Linked List class
	Appends a new node at the end. This method is
	defined inside LinkedList class shown above'''
	def insertattail(self, new_data):

		# 1. Create a new node
		# 2. Put in the data
		# 3. Set next as None
		new_node = Node(new_data)

		# 4. If the Linked List is empty, then make the
		# new node as head
		if self.head is None:
			self.head = new_node
			return

		# 5. Else traverse till the last node
		last = self.head
		while (last.next):
			last = last.next

		# 6. Change the next of last node
		last.next = new_node

	# Utility function to print the linked list
	def printList(self):
		temp = self.head
		while (temp):
			print(temp.data, end=" -> ")
			temp = temp.next
		print("NULL")


# Code execution starts here
if __name__ == '__main__':

	# Start with the empty list
	llist = LinkedList()

	llist.insertAtHead(1)
	llist.insertAtHead(2)
	print("After insertion at head:", end=" ")
	llist.printList()
	print()

	llist.insertattail(4)
	llist.insertattail(5)
	print("After insertion at tail:", end=" ")
	llist.printList()
	print()

	llist.insertAfter(1, 2)
	llist.insertAfter(5, 6)
	print("After insertion at a given position:", end=" ")
	llist.printList()

File: linked_list/singly_linked2.cpp
#include <bits/stdc++.h>
using namespace std;

// https://www.geeksforgeeks.org/insertion-in-linked-list/

/**
	In this post, methods to insert a new node in the linked list are discussed. A node can be added in three ways 
    - At the front of the linked list  
    - After a given node. 
    - At the end of the linked list.

	Method by using a constructor call to reduce memory allocation work and code lines.

	Time Complexity: O(N) 
	Auxiliary Space: O(1)
 */
class node {
public:
	int data;
	node* next;

	// A constructor is called here
	node(int value)
	{

		// It automatically assigns the
		// value to the data
		data = value;

		// Next pointer is pointed to NULL
		next = NULL;
	}
};

// Function to insert an element
// at head position
void insertathead(node*& head, int val)
{
	node* n = new node(val);
	n->next = head;
	head = n;
}

// Function to insert a element
// at a specified position
void insertafter(node* head, int key, int val)
{
	node* n = new node(val);
	if (key == head->data) {
		n->next = head->next;
		head->next = n;
		return;
	}

	node* temp = head;
	while (temp->data != key) {
		temp = temp->next;
		if (temp == NULL) {
			return;
		}
	}
	n->next = temp->next;
	temp->next = n;
}

// Function to insert an
// element at the end
void insertattail(node*& head, int val)
{
	node* n = new node(val);
	if (head == NULL) {
		head = n;
		return;
	}

	node* temp = head;
	while (temp->next != NULL) {
		temp = temp->next;
	}
	temp->next = n;
}

// Function to print the
// singly linked list
void print(node*& head)
{
	node* temp = head;

	while (temp != NULL) {
		cout << temp->data << " -> ";
		temp = temp->next;
	}
	cout << "NULL" << endl;
}

// Main function
int main()
{

	// Declaring an empty linked list
	node* head = NULL;

	insertathead(head, 1);
	insertathead(head, 2);
	cout << "After insertion at head: ";
	print(head);
	cout << endl;

	insertattail(head, 4);
	insertattail(head, 5);
	cout << "After insertion at tail: ";
	print(head);
	cout << endl;

	insertafter(head, 1, 2);
	insertafter(head, 5, 6);
	cout << "After insertion at a given position: ";
	print(head);
	cout << endl;

	return 0;
}